### PR TITLE
DatabaseCatalog drops tables faster

### PR DESCRIPTION
**Pull request by CheSema**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
DatabaseCatalog drops tables faster by using up to database_catalog_drop_table_concurrency threads

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 9561d30aaf50b1f3e444fe6e20d2806e728d9941 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66065/9561d30aaf50b1f3e444fe6e20d2806e728d9941/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66065/9561d30aaf50b1f3e444fe6e20d2806e728d9941/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by CheSema** [comment]
Stress tests (tsan) -- fatal as here https://github.com/ClickHouse/ClickHouse/issues/64945

Stateless tests (release) --
00992_system_parts_race_condition_zookeeper_long  https://github.com/ClickHouse/ClickHouse/issues/57418
03144_parallel_alter_add_drop_column_zookeeper_on_steroids -- flaks in master

Stress test (debug)
some distributed query from `00159_parallel_formatting_http.sh` hanged.

Stateless Tests (tsan)
00002_log_and_exception_messages_formatting -- noise from `Have {} tables in queue to drop. Some of them are being dropped in sync mode. Schedule background task ASAP`

**Review by tavplubix — APPROVED** [review]

File: src/Interpreters/DatabaseCatalog.cpp [hunk 1]
```diff
@@ -1218,91 +1222,168 @@ void DatabaseCatalog::dequeueDroppedTableCleanup(StorageID table_id)
...
-            dropTableFinally(table);
-            std::lock_guard lock(tables_marked_dropped_mutex);
-            [[maybe_unused]] auto removed = tables_marked_dropped_ids.erase(table.table_id.uuid);
-            assert(removed);
-        }
-        catch (...)
+        bool in_use = it->table && !it->table.unique();
+        bool old_enough = it->drop_time <= current_time;
+        if (!in_use && old_enough)
+            result.emplace_back(it);
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Interpreters/DatabaseCatalog.cpp:1260]: We could, but I was a big confused that we did all that in one iteration like it has a big sense.
I broke it into several iterations to make sure that there is no need to do it on that way.
First iteration -- just take numbers for INFO level logging. That logs are important. 
Second iteration -- collect what to drop.
Third iteration -- seeing actual data after collected tables are dropped we decide when it has to be rescheduled.

File: src/Interpreters/DatabaseCatalog.cpp [hunk 2]
```diff
@@ -833,6 +837,7 @@ DatabaseCatalog::DatabaseCatalog(ContextMutablePtr global_context_)
     , loading_dependencies{"LoadingDeps"}
     , view_dependencies{"ViewDeps"}
     , log(getLogger("DatabaseCatalog"))
+    , limitedLog(std::make_shared<LogSeriesLimiter>(log, 1, 5))
```
Comment (CheSema) [inline src/Interpreters/DatabaseCatalog.cpp:840]:
```suggestion
    , limitedLog(std::make_shared<LogSeriesLimiter>(log, 1, 20))
```

**Comment by CheSema** [comment]
Stateless Tests (debug, S3 Storage) -- `Logical error: 'Inconsistent AST formatting`
https://github.com/ClickHouse/ClickHouse/issues/62975

Stateless Tests (debug) [2/2] -- there no expected logs, strange

**Comment by azat** [comment]
This approach does not allows to scale DROP queries properly, it can only remove tables concurrently if the were lucky and end up in one batch

**Comment by clickgapai** [comment]
Hi @CheSema @tavplubix — while reviewing this PR I found the following:

- **Bug (P1):** Race condition: `getTablesToDrop` returns iterators after releasing lock, `undropTable` can invalidate them — https://github.com/ClickHouse/ClickHouse/pull/104904

Happy to discuss — close anything that's wrong or already addressed.

**Comment by clickgapai** [comment]
Update on the earlier review: CI on the proof PR #104904 confirmed the suspected bug via `Stateless tests (arm_binary, sequential)`. Filed as a separate Issue so it isn't lost when the proof PR closes:

- Race condition: `getTablesToDrop` returns iterators after releasing lock, `undropTable` can invalidate them — https://github.com/ClickHouse/ClickHouse/issues/104914

_(Automated follow-up. The proof PR has been closed; please track the bug on the Issue.)_

**Pull request merged**